Fixes a dominance violation bug in `TritonGPURemoveLayoutConversionsPass`.

levendlee opened the pull request:
In `RemoveLayoutConversion` pass, there is a step to host `triton.convert_layout` above `arith.extf`/`arith.broadcast` to reduce the memory footprint it operates on. 

1. In `hoistConvertOnTopOfExtOrBroadcast` function, the hoisted `triton.convert_layout` op is inserted right before `arith.extf`/`arith.broadcast`.

2. In `rewriteSlice` function, the rematerialized operations are also inserted right before the original operaitons, which includes the `arith.extf`/`arith.broadcast` ones.

Since step 1 is performed before step 2, the hoisted `triton.convert_layout` might use the result of rematerialized operation that is placed after it, causing dominance violation. 

Simply `insertAfter` instead of `insertBefore` in step2 resolve the above issue.

ThomasRaoux commented:
I don't think that's the correct fix as the two steps should be independent and both generate correct IR no matter what is the original IR. It sounds like the second step is incorrect but I don't understand how that could be the case. When hoisting `triton.convert_layout` it should use the source of the extf which itself should dominate the `extf` op. 
Could you share some IR exposing the problem?

ThomasRaoux commented:
> I don't think that's the correct fix as the two steps should be independent and both generate correct IR no matter what is the original IR. It sounds like the second step is incorrect but I don't understand how that could be the case. When hoisting `triton.convert_layout` it should use the source of the extf which itself should dominate the `extf` op. Could you share some IR exposing the problem?

scratch this, I think I read too fast. I think this might be the right fix but we need a test for this otherwise this would easily regress.

levendlee commented:
Sorry, I tried to create a lit unit test in a standalone fashion, it is challenging for me and I cannot get it done in half a day. Would a python integration test be acceptable for integration?

ThomasRaoux commented:
> Sorry, I tried to create a lit unit test in a standalone fashion, it is challenging for me and I cannot get it done in half a day. Would a python integration test be acceptable for integration?

Any kind of repro would be good to start with to understand the bug. It shouldn’t be very hard to extract a lit test from a python repro.

levendlee commented:
Added a lit test case.

I also changed `remat` test in `test/TritonGPU/combine.mlir` as the operands of `arith.add` are identical and the older can flip.

ThomasRaoux commented:
Thanks for including a repro, I understand the problem now. It happens in case the source Value of the extf has multiple uses and multiple of those uses are in the slice. If one of the user is above the extf the function may be doing the wrong thing.
That being said I don't think this is the correct fix as I'm pretty sure it only solves the problem "by luck" in this one case.

The right fix would be to only replace one use of extf source and not change the others. I think we should do something like this:
https://github.com/ThomasRaoux/triton/commit/5713f5510ee09c84ca312590cf2f0a9194228e81

I can try to clean this up at some point and send a PR, we should probably reduce the test as well.

levendlee commented:
Thanks so much for the fix. I am going to close this PR and wait for the proper fix to come in.

Out of curiosity, I dig more into the failure case, and finally understood the trigger of the bug. There is a special pattern inside the function that looks like the following:

```
%1 = `tt.extf` %0
%2 = `tt.convert_layout` % 1
%3 = `tt.add` %1 %1
%4 = `tt.convert_layout` %3
```

After hoist the first `tt.convert_layout`, it looks like the following

```
%1 = `tt.convert_layout` %0 # The 1st hoisted op
%2 = `tt.extf` %1 # The 1st sinked op
%3 = `tt.extf` %0 # The old op
%4 = `tt.add` %3 %3
%5 = `tt.convert_layout` %4
```

When hoist the second `tt.convert_layout`, we are facing with the situation that `%0` now has other users besides the original `tt.extf`, and the bug is triggered.

Without the proposed fix, it will looks like

```
%1 = `tt.convert_layout` %3 # The 1st hoisted op. Bug triggered!
%2 = `tt.extf` %1 # The 1st sinked op
%3 = `tt.convert_layout` %0 # The 2nd hoisted op.
%4 = `tt.extf` %3 # The 2nd sinked op
%5 = `tt.add` %4 %4
```

The repeated rewrites, especially rewriting `tt.convert_layout` for `tt.convert_layout` while the previous `tt.convert_layout` is already rewritten, makes it quite complicated for debugging. Maybe a simplified solution could be easier to maintain.

Nevertheless, thanks so much for the catch! It is a very interesting debugging experience.

gflegar commented:
Has there been any update on this? Is there an issue / PR that we can track to know when this is implemented properly, or has it already been resolved? (We are currently carrying a patch in our internal version of Triton, so would like to have a reference that can signal when this has been resolved.)

jlebar commented:
I don't believe this has been fixed.

It sounds like Thomas had a proposed fix -- could Google clean that up and send a PR?

> The right fix would be to only replace one use of extf source and not change the others. I think we should do something like this: https://github.com/ThomasRaoux/triton/commit/5713f5510ee09c84ca312590cf2f0a9194228e81

levendlee commented:
I believe this problem is resolved at https://github.com/openai/triton/pull/2722. 
I suggest Alphabet to remove the internal patch.
